beacon_node/lighthouse_network/src/discovery/mod.rs

```diff
@@ -958,12 +958,15 @@ impl<TSpec: EthSpec> NetworkBehaviour for Discovery<TSpec> {
         &mut self,
         peer_id: Option<PeerId>,
         _handler: Self::ProtocolsHandler,
-        _error: &DialError,
+        error: &DialError,
     ) {
         if let Some(peer_id) = peer_id {
-            // set peer as disconnected in discovery DHT
-            debug!(self.log, "Marking peer disconnected in DHT"; "peer_id" => %peer_id);
-            self.disconnect_peer(&peer_id);
+            // Ignore connection limit errors
+            if !matches!(error, DialError::ConnectionLimit(_)) {
+                // set peer as disconnected in discovery DHT
+                debug!(self.log, "Marking peer disconnected in DHT"; "peer_id" => %peer_id);
+                self.disconnect_peer(&peer_id);
+            }
```

Comment:
I was sondering if we should include other errors.

DialError has the following variants:
- `Banned`: Definitely we don't want to use this peer.
- `ConnectionLimit`: Not the peer's fault.
- `LocalPeerId`: A bit odd, we probably don't want to use this one.
- `NoAddresses`: Uncontactable peer.
- `DialPeerConditionFalse`: Conditions are about them being already dialed or not currently connected. None of those indicate the peer is disconnected (from the network) Not sure if we want to mark as disconnected too to prevent further dials but in principle this one does not indicate an uncontactable peer.
- `Aborted`: Docs are unclear, but looking into the code I think this happens on shutdown. Not indicative of an uncontactable peer.
- `InvalidPeerId`: Uncontable peer.
- `ConnectionIo`: Uncontable peer.
- `Transport`: Uncontable peer.

So I think here it would be better to match the error and explicitly decide which ones we take as disconnected, and include too `DialPeerConditionFalse` and `Aborted` as whitelisted errors.